text/0000-debug-map-key-value.md

```diff
@@ -0,0 +1,284 @@
+            } else {
+                if self.has_fields {
+                    self.fmt.write_str(", ")?
+                }
+                key.fmt(self.fmt)?;
+            }
+
+            // Mark that we're in an entry
+            self.has_key = true;
+            Ok(())
+        });
+
+        self
+    }
+
+    pub fn value(&mut self, value: &dyn fmt::Debug) -> &mut DebugMap<'a, 'b> {
+        self.result = self.result.and_then(|_| {
+            // Make sure there is a partial entry to finish
+            if !self.has_key {
+                return Err(fmt::Error);
+            }
+
+            if self.is_pretty() {
+                let mut slot = None;
+                let mut writer = PadAdapter::wrap(&mut self.fmt, &mut slot);
+                writer.write_str(": ")?;
+                value.fmt(&mut writer)?;
+                writer.write_str(",\n")?;
+            } else {
+                self.fmt.write_str(": ")?;
+                value.fmt(self.fmt)?;
+            }
+
+            // Mark that we're not in an entry
+            self.has_key = false;
+            Ok(())
+        });
+
+        self.has_fields = true;
+        self
+    }
+
+    pub fn finish(&mut self) -> fmt::Result {
+        // Make sure there isn't a partial entry
+        if self.has_key {
+            return Err(fmt::Error);
+        }
+
+        self.result.and_then(|_| self.fmt.write_str("}"))
+    }
+}
+```
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+The proposed `key` and `value` methods are't immediately useful for `Debug` implementors that are able to call `entry` instead. This creates a decision point where there wasn't one before. The proposed implementation is also going to be less efficient than the one that exists now because it introduces a few conditionals.
+
+On balance, the additional `key` and `value` methods are a small and unsurprising addition that enables a set of use-cases that weren't possible before, and aligns more closely with `serde`.
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+The universal alternative of simply _not doing this_ leaves consumers that do need to format map keys independently of values with a few options:
+
+- Write an alternative implementation of the format builders. The output from this alternative implementation would need to be kept reasonably in-sync with the one in the standard library. It doesn't change very frequently, but does from time to time. It would also have to take the same care as the standard library implementation to retain formatting flags when working with entries.
+- Buffer keys and format them together with values when the whole entry is available. Unless the key is guaranteed to live until the value is supplied (meaning it probably needs to be `'static`) then the key will need to be formatted into a string first. This means allocating (though the cost could be amortized over the whole map) and potentially losing formatting flags when buffering.
+
```

Comment:
A third option here is to simply not error when keys and values come alone. That might actually be useful in some weird semi-map semi-array cases. That would also be more efficient?

Reply:
Hmm, that's an interesting thought 🤔 I think that might just mask incorrect implementations though. I imagine we'd be tracking the same additional state too so that we knew when a key or value called 'out of order' should be formatted in a set-like way rather than a map-like way.

Reply:
Yeah quite possibly.
Would be good to record this in the text of the RFC in any case. :)